Libhal2.0 port

Coreyboy1820 opened the pull request:
Finished:
- Speed sensor abstract base class
- Speed sensor adaptor from hal::drc
- Mission control abstract base class
- esp8266 implementation of mission control interface
- transitioning homing function to work on a span of servos and magnets.
- Updated all implementations in arm and drive to use the abstract base classes instead of hard set implementations
- Now using libhal2.0 (pre-built binaries)
- Created an offset servo implementation of servo for homing purposes.
- templated the homing function

Coreyboy1820 committed:
:construction: working on porting to 2.0, started base file structure

Coreyboy1820 committed:
started porting to libhal prebuilts

Coreyboy1820 committed:
working on port

Coreyboy1820 committed:
Nearly finished up the init platform. Will work on mc next

Coreyboy1820 committed:
:sparkles: finished the initalizing platform

Coreyboy1820 committed:
:sparkles: :construction: finished converting all the basic logic to libhal2.0

Coreyboy1820 committed:
:constructions: nealy finished mission control implementation

Coreyboy1820 committed:
:construction: finished adding mission control just need to fix errors

Coreyboy1820 committed:
This is before I switch all of the pairs to using pointers

Coreyboy1820 committed:
:sparkles: finished switching the servos over to offset homing

Coreyboy1820 committed:
:sparkles: finished changing from shared pointers to regular pointers, then allowing for multiple legs to be used in the control loop

Coreyboy1820 committed:
updated the mission control

Coreyboy1820 committed:
:sparkles: changed homing function to be a template function so it no longer needs pairs when passing in legs.

Coreyboy1820 committed:
:sparkles: switched away from int to size_t

Coreyboy1820 committed:
:sparkles: finished switching from templates

Coreyboy1820 committed:
:sparkles:updated code to use homing structs instead of passing in seperate magnets, homed booleans, and servos while matching them

Coreyboy1820 committed:
:bug:finished testing and slight bug fixing

Coreyboy1820 committed:
fixed arm to build succesfully

Viha123 committed:
untested stuff to get commands

Viha123 committed:
fixing bugs drive esp

Coreyboy1820 committed:
:bandaid: fixed project name in cmake

Coreyboy1820 committed:
:construction: finished testing for the day

Coreyboy1820 committed:
:construction: finished testing for the day

Coreyboy1820 committed:
Fixed main part of code

Coreyboy1820 committed:
Fixed main part of code

Coreyboy1820 committed:
found issue with version numbers

Coreyboy1820 committed:
found issue with version numbers

kammce commented on drive/platform-implementations/esp8266_mission_control.cpp:
```diff
@@ -55,71 +56,56 @@ class esp8266_mission_control : public mission_control
+
+    if (!m_header_finished &&
+        hal::finished(m_http_header_parser.find_end_of_header)) {
+      auto content_length = m_http_header_parser.parse_content_length.value();
```
```suggestion
      auto& content_length = m_http_header_parser.parse_content_length.value();
```

If you make this change this should start working. auto doesn't copy reference modifiers over. So you have to explicitly make this a reference. 

Value() returns a reference to the internal variable. stream_fill takes a reference/pointer to an integer (honestly bad design but it was meant for handling the single pipeline case without breaking the pipeline). I'll deciding on how to change the API without breaking everyone else's code.

Coreyboy1820 committed:
:construction: :sparkles: got code to cycle through but the floats are outputting weird numbers. Need to test in person

Coreyboy1820 committed:
:construction: :sparkles: got code to cycle through but the floats are outputting weird numbers. Need to test in person

Coreyboy1820 committed:
:sparkles: reverted back to using hardware and not print implementations

Coreyboy1820 committed:
:sparkles: reverted back to using hardware and not print implementations

Coreyboy1820 committed:
:sparkles: added floating point support in cmake

Coreyboy1820 committed:
:sparkles: added floating point support in cmake

Coreyboy1820 committed:
Merge branch 'libhal2.0-port-reverting-to-hardware' into libhal2.0-port

FluffyFTW committed:
builds lmao

Coreyboy1820 committed:
Parsing commands currently works, just need to work out a bug which closes the connection

Coreyboy1820 committed:
Parsing commands currently works, just need to work out a bug which closes the connection

FluffyFTW committed:
buildable

Coreyboy1820 committed:
:alien: Libhal 2.0 update for arm

- Arm switched from drc to mc_x
- updated esp8266_mission_control.cpp to maintain content length

Coreyboy1820 committed:
.

Coreyboy1820 committed:
.

FluffyFTW committed:
angles change

Coreyboy1820 committed:
bug found where subtracting in offset servo instead of adding

Coreyboy1820 committed:
bug found where subtracting in offset servo instead of adding

Coreyboy1820 committed:
debugging offsets with servos

FluffyFTW committed:
tmp

FluffyFTW committed:
building debug on ubuntu22.04

Coreyboy1820 committed:
changed order of legs in drive

FluffyFTW committed:
changed max rmp to 2

FluffyFTW committed:
esp not updating

FluffyFTW committed:
10/16

FluffyFTW committed:
10/17

Coreyboy1820 committed:
added static to socket config in lpc

FluffyFTW committed:
10/17

FluffyFTW committed:
esp works now

FluffyFTW committed:
fuck lerp

Coreyboy1820 committed:
Merge branch 'libhal2.0-arm' of https://github.com/SJSURoboticsTeam/urc-control-system-2023 into libhal2.0-arm

Coreyboy1820 committed:
:alien: merged arm libhal2.0 port into main 2.0 port

Coreyboy1820 committed:
:alien: ported arms esp over to drives

Coreyboy1820 committed:
:alien: Updated to libhal 2.0

- arm upgraded to 2.0
- drive upgraded to 2.0
- mission control uses new esp8266

Coreyboy1820 committed:
:construction: fixed bugs, now need to revert parse commands

Coreyboy1820 committed:
:sparkles: brought in old drive parse commands and fixed it, code now compiles

FluffyFTW committed:
:ambulance: Fixed homing

- Fixed homing issue
- Fixed parsing changes
- Imported arm esp algorithm

Coreyboy1820 committed:
:sparkles: fully finished converting arm and drive to libhal 2.0, arm was working but I slowed down the rate at which we querried the server and decreaded the delay in arm from 10->6ms

Coreyboy1820 committed:
:fire: removed all of the old code

kammce commented on arm/applications/application.cpp:
```diff
@@ -44,15 +45,18 @@ hal::status application(sjsu::arm::application_framework& p_framework)
-    auto timeout = hal::create_timeout(clock, 5s);
-    commands = mission_control.get_command(timeout).value();
+    if(loop_count==10) {
+      auto timeout = hal::create_timeout(clock, 5s);
```
```suggestion
      auto timeout = hal::create_timeout(clock, 100ms);
```

You do not want to deadlock the rover for 5s while you figure out whether or not you are connected. This time is too long and if the rover is no-responsive could harm someone. We do have the E-stop but this time should be reduced. In this configuration, you'll reach this every 60ms (good choice, I think my best with ExpressJS was 50ms or so), then an additional 100ms. So every reconnect attempt happens around every 160ms. Once we get this pass in, we'll need to go over how to improve how timing is done for this control loop. It makes more sense to push the frequency of checks INTO the mission control driver itself rather than having to manage it outside. But this was my suggestion as a quick fix.

kammce commented on drive/applications/application.cpp:
```diff
@@ -37,10 +38,13 @@ hal::status application(application_framework& p_framework)
 
   while (true) {
+    if(loop_count==10) {
+      auto timeout = hal::create_timeout(clock, 5s);
```
ditto here

Coreyboy1820 committed:
:sparkles::sparkles: Finished the port to libhal2.0, everything on arm and drive is functioning

Coreyboy1820 committed:
updating pipeline

Coreyboy1820 committed:
handled merges

Coreyboy1820 committed:
testing build

Coreyboy1820 committed:
testing build
